### PR TITLE
BaSP-MR-47: ASYNC AWAIT and fix issues

### PR DESCRIPTION
**Pull request by flordimonaco**

### CODE REVIEW

File: src/controllers/subscription.js [hunk 1]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
   const { classes, member, date } = req.body;
 
-  if (!mongoose.Types.ObjectId.isValid(id)) {
+  if (!mongoose.isValidObjectId(id)
+  || (classes && !mongoose.isValidObjectId(classes))
+  || (member && !mongoose.isValidObjectId(member))) {
     return res.status(400).json({
       error: true,
-      message: 'ID is not valid',
+      message: 'ID or class or member ID is not valid',
```
Comment (DelfiLT) [inline src/controllers/subscription.js:118]: I think the second message it's a little bit confusing with so many or, you could say something like "One of the IDs is invalid" or "Subscription, Class and Member IDs must have a valid ID"

**Review by DelfiLT — APPROVED** [review]

File: src/controllers/subscription.js [hunk 2]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
+        message: 'Subscription not found',
+        data: undefined,
       });
-    })
-    .catch((error) => res.status(500).json({
+    }
+
+    return res.status(200).json({
+      error: false,
+      message: `Subscription found: ${foundSubscriptionById}`,
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:46]: Could u please remove the `foundSubscriptionById` to the message?

File: src/controllers/subscription.js [hunk 3]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
-  Subscription.create({
-    classes,
-    member,
-    date,
-  })
-    .then((newSubscription) => res.status(201).json({
+
+  if (!mongoose.isValidObjectId(classes) || !mongoose.isValidObjectId(member)) {
+    return res.status(400).json({
+      error: true,
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:63]: Please add the data attribute with undefined value

File: src/controllers/subscription.js [hunk 4]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
+    });
+  }
+
+  try {
+    const existingClass = await Class.findById(classes);
+    const existingMember = await Member.findById(member);
+
+    if (!existingClass || !existingMember) {
+      return res.status(404).json({
+        error: true,
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:74]: Same here with data

File: src/controllers/subscription.js [hunk 5]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
+      });
+    }
+
+    const existingSubscription = await Subscription.findOne({ classes, member, date });
+
+    if (existingSubscription) {
+      return res.status(400).json({
+        error: true,
+        message: 'Subscription already exists',
+        data: existingSubscription,
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:85]: Here data value should be `undefined`

File: src/controllers/subscription.js [hunk 6]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
-    .catch((error) => res.status(500).json({
+    }
+
+    const existingSubscription = await Subscription.findOne({ classes, member, date });
+
+    if (existingSubscription && existingSubscription.id.toString() !== id) {
+      return res.status(400).json({
+        error: true,
+        message: 'Subscription already exists',
+        data: existingSubscription,
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:139]: Same here with data: undefined

File: src/controllers/subscription.js [hunk 7]
```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
...
+      id,
+      {
+        classes,
+        member,
+        date,
+      },
+      { new: true },
+    ).populate('classes member');
+
+    if (!updatedSubscription) {
```
Comment (iaracriscenti) [inline src/controllers/subscription.js:153]: You could check if the subscription with the id is found at the beginning of the controller by using the findById method, and then you run the rest of the validations, wdyt?

**Review by iaracriscenti — CHANGES_REQUESTED** [review]
Great additions! Just left a few things to check

**Review by iaracriscenti — APPROVED** [review]
🚀

**Pull request merged**